### PR TITLE
[DataGrid] Prepare the tree structure for grouping sorting / filtering + multi field grouping

### PR DESCRIPTION
**Pull request by flaviendelangle**
Needs to be merged before next release

Extracted from #3277 
Breaking compared to #2725 but not breaking compared to any released version.

- [x] Replace `rowNode.groupingValue` with `rowNode.groupingKey` and `rowNode.groupingField`. I was relying on the depth, but for the grouping columns I want to handle unbalanced groups so I need to store both the key and the field (see https://www.ag-grid.com/javascript-data-grid/grouping-unbalanced-groups/). The goal is to differentiate two grouping criteria being at the same depth but from a different field. For instance when grouping `[{ director: 'A', company: 'B' }, { director: null, company: 'A'  }` by `['director', 'company']`, both will have `A` as the grouping key of depth 0 but not from the same field.

- [x] Pass the `groupingKey` on the `valueGetter` of the grouping cell to prepare for sorting / filtering and access directly the `rowNode` on the component for rendering

- [x] Reduce the depth-margin on the grouping cell based on a feedback of @oliviertassinari (the 3 Argos diffs come from there)

### CODE REVIEW

**Review by m4theushw — APPROVED** [review]

File: docs/src/pages/components/data-grid/group-pivot/CustomGroupingColumnTreeData.js [hunk 1]
```diff
@@ -60,13 +72,53 @@ CustomGridTreeDataGroupingCell.propTypes = {
    */
   id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
   /**
-   * The cell value, but if the column has valueGetter, use getValue.
+   * The node of the row that the current cell belongs to.
    */
-  value: PropTypes.shape({
+  rowNode: PropTypes.shape({
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/src/pages/components/data-grid/group-pivot/CustomGroupingColumnTreeData.js:77]: Done :+1:

**Comment by alexfauquette** [comment]
There is one point I do not understand. In the code you modified, you replace `path=["A", "A"]` by `path = [
            { key: 'A', field: null },
            { key: 'A', field: null },
          ]`. But when I look at Demo examples, they use `hierarchy` to define how they are grouped. Is there a link between `path` and `hierarchy` or is it two distinct things?

File: docs/src/pages/components/data-grid/group-pivot/CustomGroupingColumnTreeData.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/src/pages/components/data-grid/group-pivot/CustomGroupingColumnTreeData.js]: @m4theushw those transpiled files do not apply the `generateProptypes` config but another one stored on `formattedTSDemos`

I will do, a follow up PR trying to unify both :+1:

**Comment by flaviendelangle** [comment]
@alexfauquette `buildRowTree` is a method called by `useGridTreeData` to create the tree from the rows.
It will also be used by `useGroupingColumns`.

Those hooks pre-process the data to transform them into the `path` format `buildRowTree` expects.

For `useGridTreeData`, it is simply calling `getTreeDataPath` and setting the `field` to `null` 

```ts
      const rows = params.ids
        .map((rowId) => ({
          id: rowId,
          path: props.getTreeDataPath!(params.idRowsLookup[rowId]).map(
            (key): BuildRowTreeGroupingCriteria => ({ key, field: null }),
          ),
        }))
        .sort((a, b) => a.path.length - b.path.length);
```

For `useGroupingColumns` it is slightly more complicated because we must 1st extract the distinct values of each grouping column.


```ts
      params.ids.forEach((rowId) => {
        const row = params.idRowsLookup[rowId];

        groupingColumnsModel.forEach((groupedByField) => {
          const { key } = getCellGroupingCriteria({
            row,
            id: rowId,
            colDef: columnsLookup[groupedByField],
            field: groupedByField,
          });
          const groupingFieldsDistinctKeys = distinctValues[groupedByField];

          if (key != null && !groupingFieldsDistinctKeys.map[key.toString()]) {
            groupingFieldsDistinctKeys.map[key.toString()] = true;
            groupingFieldsDistinctKeys.list.push(key);
          }
        });
      });

      const rows = params.ids.map((rowId) => {
        const row = params.idRowsLookup[rowId];
        const parentPath = groupingColumnsModel
          .map((groupingField) =>
            getCellGroupingCriteria({
              row,
              id: rowId,
              colDef: columnsLookup[groupingField],
              field: groupingField,
            }),
          )
          .filter((cell) => cell.key != null);

        return {
          path: [...parentPath, { key: rowId.toString(), field: null }],
          id: rowId,
        };
      });
```

**Pull request merged**